kv/migration.go

```diff
@@ -0,0 +1,332 @@
+type MigrationFunc func(context.Context, Store) error
+
+// AnonymousMigration is a utility type for creating migrations from anonyomous functions.
+type AnonymousMigration struct {
+	name string
+	up   MigrationFunc
+	down MigrationFunc
+}
+
+// NewAnonymousMigration constructs a new migration from a name string and an up and a down function.
+func NewAnonymousMigration(name string, up, down MigrationFunc) AnonymousMigration {
+	return AnonymousMigration{name, up, down}
+}
+
+// Name returns the name of the migration.
+func (a AnonymousMigration) Name() string { return a.name }
+
+// Up calls the underlying up migration func.
+func (a AnonymousMigration) Up(ctx context.Context, store Store) error { return a.up(ctx, store) }
+
+// Down calls the underlying down migration func.
+func (a AnonymousMigration) Down(ctx context.Context, store Store) error { return a.down(ctx, store) }
+
+// Migrator is a type which manages migrations.
+// It takes a list of migration specifications and undo (down) all or apply (up) outstanding migrations.
+// It records the state of the world in store under the migrations bucket.
+type Migrator struct {
+	logger     *zap.Logger
+	Migrations []MigrationSpec
+
+	now func() time.Time
+}
+
+// NewMigrator constructs and configures a new Migrator.
+func NewMigrator(logger *zap.Logger, ms ...MigrationSpec) *Migrator {
+	m := &Migrator{
+		logger: logger,
+		now: func() time.Time {
+			return time.Now().UTC()
+		},
+	}
+	m.AddMigrations(ms...)
+	return m
+}
+
+// AddMigrations appends the provided migration specs onto the Migrator.
+func (m *Migrator) AddMigrations(ms ...MigrationSpec) {
+	m.Migrations = append(m.Migrations, ms...)
+}
+
+// Initialize creates the migration bucket if it does not yet exist.
+func (m *Migrator) Initialize(ctx context.Context, store Store) error {
+	return store.Update(ctx, func(tx Tx) error {
+		_, err := tx.Bucket(migrationBucket)
+		return err
+	})
+}
+
+// List returns a list of migrations and their states within the provided store.
+func (m *Migrator) List(ctx context.Context, store Store) (migrations []Migration, _ error) {
+	if err := m.walk(ctx, store, func(id influxdb.ID, m Migration) {
+		migrations = append(migrations, m)
+	}); err != nil {
+		return nil, err
+	}
+
+	migrationsLen := len(migrations)
+	for idx, migration := range m.Migrations[migrationsLen:] {
+		migration := Migration{
+			ID:   influxdb.ID(migrationsLen + idx + 1),
+			Name: migration.Name(),
+		}
+
+		migrations = append(migrations, migration)
+	}
+
+	return
+}
+
+// Up applies each outstanding migration in order.
+// Migrations are applied in order from the lowest indexed migration in a down state.
+//
+// For example, given:
+// 0001 add bucket foo         | (up)
+// 0002 add bucket bar         | (down)
+// 0003 add index "foo on baz" | (down)
+//
+// Up would apply migration 0002 and then 0003.
+func (m *Migrator) Up(ctx context.Context, store Store) (err error) {
+	defer func() {
```

Comment:
Similar situation here. I'm not sure if there is a better way, but these assignments in `defer` always confuse me at first read.